docs/proposal-new-output-syntax.md

```diff
@@ -0,0 +1,211 @@
+
+- `a:x => b` means `x` is an *expected output* of `a`
+  - if `a` generates output `x` we spawn `b`
+  - if `a` finishes without generating `x` we do not spawn `b`
+    - but we retain `a` in the n=0 task pool as an *incomplete task*, and flag
+      that as a problem
+- `a?:x => b` means `x` is an *optional output* of `a`
+  - if `a` generates output `x` we spawn `b`
+  - if `a` finishes without generating `x` we do not spawn `b`
+    - (but we do not retain `a` as an incomplete task)
+
+Abbreviated syntax for the success case:
+- `a => b` or `a: => b` means `a:succeed => b` (success is expected)
+- `a? => b` means `a?succeed => b` (success is optional)
+
+## Differences from the Proposed Optional Trigger Syntax
+
+Consider `a:x => b` where `a` completes without generating `a:x`.
+
+- With expected triggers: we remove `a` as completed but spawn `b` into the n=0
+  pool anyway as an *unsatisfied waiting task*, and flag that as a problem
+  - the user has to figure out why `b` is waiting (and the solution likely
+    involves fixing and rerunning the upstream task `a`)
+
+- With expected outputs: we do not spawn `b` but we retain `a` in the n=0 pool
+  as an *incomplete task*, and flag that as a problem
+  - the user can see `a` did not generate an expected output (and the solution
+    likely involves messing with the flagged task `a`)
+  - we do not spawn waiting children that (depending on the resolution of the
+    problem) might not be needed
+
+## Path Branching 
+
+Concurrent paths that trigger off of non mutually exclusive expected outputs:
+```
+a:x => b1  # concurrent path branching
+a:y => b2
+b1 & b2 => c  # ... joining
+```
+
+Alternate paths that trigger off of optional mutually exclusive outputs:
+```
+a?x => b1  # alternate path branching
+a?y => b2
+b1 | b2 => c  # ... joining
+```
+
+Failure recovery case:
+```
+a? => b1  # alternate path branching
+a?fail => b2
+b1 | b2 => c  # ... joining
+```
+
+Graph hack to require that at least one alternate path runs:
+```
+a?x => b1  # alternate path branching
+a?y => b2
+b1 | b2 => c  # ... joining
+a => c  # HACK
+```
+Here, `c` will be spawned if `a` succeeds, but will be stuck as waiting if
+neither b1 nor b2 succeed. Note the same graph hack is required for optional
+triggers. But we could avoid the hack with something like this:
+```
+[runtime]
+   [[a]]
+       outputs completion = "x|y"  # NEW
+       [[[outputs]]]
+          x = x
+          y = y
+```
+
+## No Special Treatment of Failed Tasks Needed
+
+Unhandled failed case:
+```
+a => b  # success expected
+```
+If `a` fails it automatically gets retained in the `n=0` pool as an incomplete
+task (and will show as failed as well, in this case).
+
+Handled failed case:
+```
+a? => b
+a?fail => c
+```
+If `a` fails no problem, success was optional.
+
+Bonus, no additional special syntax (e.g. `c:fail => NULL` or `c:fail => !c`) is needed to handle leaf tasks:
+```
+a => b => c?  # success of c is optional
```

Comment:
Well, it is sort of special syntax because RHS triggers usually don't mean anything (I didn't realise they were allowed).

What's the logic going to be? Success is assumed to be required unless there is a `succeed` or `fail` trigger?

```
a => c
b => c? # Success of c is optional
# Also allowed:
b => c?succeed
b => c?fail
```
```
a => b
b?fail => c # Success of b is optional
```
```
a => b
b:fail => c # Failure of b is required!
```

Reply:
Yes, `c?` is shorthand for `c?succeeded` similar to `c` being shorthand for `c:succeeded`.

Reply:
@dpmatthews remember `?succeed` is an output, not a trigger.   So `b => c?` means: `c triggers off of b; then when c runs, success is not a required output* (so if it does not succeed, we don't have to keep it in n=0 as an incomplete/blocked task).

Reply:
> ```
> a => b
> b:fail => c # Failure of b is required!
```

Yes, fail triggers should almost always be this:
```
b?fail => c  # failure is "optional" but if it happens trigger c
```
But note will now be able to have tasks that are expected to fail - which might occasionally be useful!

Reply:
> we don't have to keep it in n=0 as an incomplete/blocked task

I think Dave would like to keep the tasks in n=0 to make it easier for the user to fix the issue (retrigger, trigger-edit, artificially satisfy outputs)?

Reply:
In that case, use `b => c` (and keep `c` in n=0 if it fails) instead of `b => c?`.

Reply:
Instead of "keep unhandled failed tasks in n=0" we now have "keep incomplete tasks in n=0", which automatically includes failed tasks whose success is required (`b => c`)

Reply:
> `?succeed` is an output, not a trigger. So `b => c?` means: _c triggers off of b; then when c runs, success is not a required output_ (so if it does not succeed, we don't have to keep it in n=0 as an incomplete/blocked task).

Yes, sorry - didn't explain myself very well. Presumably you agee that `a => b` doesn't mean that success of `b` is required (even though `b` is shorthand for `b:succeed`)? See examples above.

If there is neither a `fail` output nor a `succeed` output defined then `succeed` is assumed to be required.

What about `a => b:succeed`. Does that make success of `b` required? Presumably yes (although I can't see why you'd need to use this).

Reply:
Sorry, I'm confused about what you're saying here exactly. 

Does this statement:
> Presumably you agree that `a => b` doesn't mean that success of `b` is required?

not contradict this one?:
> If there is neither a `fail` output nor a `succeed` output defined then `succeed` is assumed to be required.

I think `a => b` **does** imply success of `b` is required.  That's exactly why we can keep the failed `b` in n=0 as an incomplete task.

Reply:
Maybe you're worried about this:
```
a => b
b:fail => c # Failure of b is required!
```
which says both success and failure of b are required.   That's an error because `:succeed` and `:fail` are mutually exclusive outputs, so we should make it fail validation.

Reply:
So, if I understand correctly, I assume that means this is an error:

```
a => b
b? => c
b?fail => d
```
i.e. `:succeed` and `?fail` are mutually exclusive outputs

So this is also an error:
```
a => b?   # Should be a => b:fail
b:fail => c
```

What about this?
```
a => c
b => c?
```
Are `:succeed` and `?succeed` mutually exclusive?

> That's an error because `:succeed` and `:fail` are mutually exclusive outputs, so we should make it fail validation.

Yes, I think it would be essential to detect graphing errors at validation (too easy to make mistakes otherwise).
However we can't do this for Cylc 7 workflows so we'd need to do something to detect Cylc 8 workflows (based on use of newer syntax?).

Reply:
> Are `:succeed` and `?succeed` mutually exclusive?

Yes, because the "output type" (for want of a better name?) refers to the **output** not to the trigger, so it is not possible for a single output to be both required and not-required.  So this is definitely an error:
```
# ERROR
a:succeed => b
a?succeed => c
```
Similarly:
```
# ERROR
a => z:succeed
b => z?succeed
```

If "output status" (optional vs required) is not specified we have two options:
1. it means success is required, or
2. it means success is required *unless

Reply:
So your first example is only an error if we `b` **always** means `b:succeed` (as opposed to: assume `:succeed` unless told otherwise):
```
a => b
b? => c
b?fail => d
```

Your second example is an error, because it doesn't make sense for failure to be required at the same time as success is optional.  Succeed and fail are mutually exclusive so either one has to be required, or both have to be optional (but you don't have to write both optional cases out if you only trigger off of one of them).

```
a => b?   # Should be a => b:fail
b:fail => c
```
And your final example is a simpler version of the first. It's an error unless we go for the less prescriptive "assume success required unless told otherwise" option:
```
a => c
b => c?
```

Reply:
The vast majority of tasks will be "success required" of course, so users really only have to use this syntax at alternate path branch points and leaf tasks where success/fail doesn't matter.

Reply:
I vote for the more flexible option: assume success is required unless told otherwise.  That way, we don't have to explicitly use a `?` every time an affected task appears at the end of a graph line.

Reply:
Whichever option we choose, I think we have to _assume success is required unless it's explicitly made optional elsewhere in the graph_ to deal with cases like `a => b:x => c` or `a => b:start =? c`. In these cases we need to assume success is required.
Also note that any of `?succeed`, `:fail`, `?fail` will make success optional.

I tend to think it will be easier if we always treat `b` as a shorthand for `b:succeed` rather than trying to explain the exception. This is likely to lead to more graphing errors but that's OK if we can fail at validation (and if we detect a Cylc 7 workflow we can still issue warnings since these will help identify where the graph needs changing).

Reply:
I don't think the proposal covers this discussion yet?
I think we have to assume `succeed` is required unless it's explicitly made optional elsewhere in the graph (via `?succeed` or `?fail`).
Similarly for `submit`.
And `a => b` implies `succeed` is expected for `b` regardless of what else is in the graph.

Reply:
> I think we have to assume succeed is required unless it's explicitly made optional elsewhere in the graph (via `?succeed` or `?fail`).
Similarly for submit.

Agreed, I'll update the document to make that clear.

>  And `a => b` implies `succeed` is expected for `b` regardless of what else is in the graph.

I don't understand this. Presumably you're not saying that `a` has to succeed in order to trigger `b`, which obviously true regardless of whether `a:succeed` is a required or optional output.  So you must be saying that `a => b` means `a => b:succeed` even if `b?succeed` or `b?fail` appear elsewhere in the graph?  But that contradicts the general rule above. :confounded:

Reply:
It doesn't contradict. The general rule only applies in cases where success or failure hasn't been specified as expected or optional. It's required for cases like `a => b:start => c` (with no other reference to `b` in the graph) - `succeed` hasn't been made expected or optional in this case for `b` so we assume it is expected.

Reply:
In the latest version you've said: "`foo` means `foo:succeed` UNLESS ..."
That's only true when `foo` is the RHS expresssion, e.g. `bar => foo`.
The "unless" doesn't apply if it's the LHS, e.g.  `foo => bar` - in this case it always means `foo:succeed`.
So, it's a bit awkward to explain.
You've also said "For complete safety, if success is optional use explicit syntax throughout."
If that's the case (and especially if we're going to give validation warnings about this) why not make it mandatory?

I'm not sure I understand the bit about supporting Cylc 7 graphs but we can handle that via the file name (`suite.rc`) if necessary.

Reply:
>  If that's the case (and especially if we're going to give validation warnings about this) why not make it mandatory?

Yep, now mandatory